favicon-from-sprite.js

```diff
@@ -0,0 +1,149 @@
+  var FRAME_DELAY = 400;
+
+  // END CONFIG
+
+  var FAVICON_SIZE = 16; // pixels
+  var ONE_PIXEL_SCALED = FAVICON_SIZE / tilesize;
+  PIXEL_PADDING *= ONE_PIXEL_SCALED;
+  var canvas = document.createElement('canvas');
+  canvas.width = FAVICON_SIZE + 2*PIXEL_PADDING;
+  canvas.height = FAVICON_SIZE + 2*PIXEL_PADDING;
+  var ctx = canvas.getContext('2d');
+  var faviconLinkElem;
+  var faviconFrameURLs = [];
+
+  var _startExportedGame = startExportedGame;
+  globals.startExportedGame = function() {
+    _startExportedGame.apply(this, arguments);
+    startFaviconLoop();
+  }
+
+  function startFaviconLoop() {
+    var frameNum = 0;
+    var frames = getFrames(SPRITE_NAME);
+
+    faviconFrameURLs = frames.map(drawFrame);
+
+    // Only one frame? Don't even bother with the loop, just paint the icon once.
+    if (frames.length === 1) {
+      return updateBrowserFavicon(faviconFrameURLs[0]);
+    }
+
+    setInterval(function() {
+      frameNum = ++frameNum % frames.length;
+      updateBrowserFavicon(faviconFrameURLs[frameNum]);
+    }, FRAME_DELAY);
+  }
+
+  function drawFrame(frameData) {
+    var pal = getPalette(PALETTE_ID);
+    var bgColor = pal && pal[BG_COLOR_NUM] || [20,20,20];
+    var spriteColor = pal && pal[FG_COLOR_NUM] || [245,245,245];
+    var rounding_offset = ROUNDED_CORNERS ? ONE_PIXEL_SCALED : 0;
+
+    // Approximate a squircle-shaped background by drawing a fat plus sign with
+    // two overlapping rects, leaving some empty pixels in the corners.
+    var longSide = FAVICON_SIZE + 2*PIXEL_PADDING;
+    var shortSide = longSide - rounding_offset*ONE_PIXEL_SCALED;
+    ctx.fillStyle = rgb(bgColor);
+    ctx.fillRect(rounding_offset,
+                 0,
+                 shortSide,
+                 longSide);
+    ctx.fillRect(0,
+                 rounding_offset,
+                 longSide,
+                 shortSide);
+
+    // Draw sprite foreground.
+    ctx.fillStyle = rgb(spriteColor);
+    for (var y in frameData) {
+      for (var x in frameData[y]) {
+        if (frameData[y][x] === 1) {
+          ctx.fillRect(x*ONE_PIXEL_SCALED + PIXEL_PADDING,
+                       y*ONE_PIXEL_SCALED + PIXEL_PADDING,
+                       ONE_PIXEL_SCALED,
+                       ONE_PIXEL_SCALED);
+        }
+      }
+    }
+
+    return canvas.toDataURL("image/x-icon");
+  }
+
+  function updateBrowserFavicon(dataURL) {
+    // Add or modify favicon link tag in document.
+    faviconLinkElem = faviconLinkElem || document.querySelector('#favicon');
+    if (!faviconLinkElem) {
+      faviconLinkElem = document.createElement('link');
+      faviconLinkElem.id = 'favicon';
+      faviconLinkElem.type = 'image/x-icon';
+      faviconLinkElem.rel = 'shortcut icon';
+      document.head.appendChild(faviconLinkElem);
+    }
+    faviconLinkElem.href = dataURL;
+  }
+
+  function getFrames(spriteName) {
+    var spriteId = names.sprite.get(spriteName) || playerId;
```

Comment:
this part doesn't work anymore (`names` must get populated later), so the favicon always ends up being the player
the `sprite` global is available and populated though; could fix this by either looping through `sprite` and comparing names, or just make the `SPRITE_NAME` option be `SPRITE_ID`

Reply:
Good catch; I forgot to test with other sprites after the refactor. I'd prefer to keep the ability to use a sprite name for usability so authors don't have to dig thru their game data for the ID. 

`names.sprite` _should_ get populated by calls to `parseSprite()`, which also populates the `sprite` object. I'll take another look at it. I'm not a huge fan of the way Bitsy stores this info; it feels a little haphazard.